### PR TITLE
Create script API for reading and writing zip files

### LINKED ISSUE #1069: Expose support for packing/unpacking ZIP archives to the scripting API
**Describe the project you are working on:**
Node based image and material editor.

**Describe the problem or limitation you are having in your project:**
I would like to be able to pack and unpack user configuration to a single zip file. I will for sure find another usage for it later.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I will be able to do what I cannot do now.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Inspiration can be taken from gdunzip (link below). Zip functionality can be a method in File class (which could zip files and directories).

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It cannot be done with a few lines of script.

**Is there a reason why this should be core and not an add-on in the asset library?:**
The plugin in GDScript was made but it's only unzip, it doesn't do CRC checks and it's not very fast. The engine already uses some C library (still minizip?) so it shouldn't be hard to expose it to GDScript to get native performance.

Related links [3701](https://github.com/godotengine/godot/issues/3701), [22440](https://github.com/godotengine/godot/issues/22440) and [gdunzip](https://github.com/jellehermsen/gdunzip).

*Bugsquad edit (keywords for easier searching): extract, create*

### PR DESCRIPTION
**Pull request by jameswestman**
*Bugsquad edit: this was salvaged and merged in https://github.com/godotengine/godot/pull/65281.*
*Bugsquad edit: `master` version of https://github.com/godotengine/godot/pull/58900.*

This builds on #22554, which only supported writing zip files.

Creates two new classes, ZipReader and ZipWriter, for reading and writing zip files in GDScript using the already bundled minizip library.

Example:
```gdscript
var p := ZIPPacker.new()
p.open("user://test.zip")
p.start_file("hello/world.txt")
p.write_file("Hello, world!".to_utf8_buffer())
p.close()

var r := ZipReader.new()
r.open("user://test.zip")
for filename in r.get_files():
    print(filename, ": ", r.read_file(filename).get_string_from_utf8())
```

Note: Append modes don't work yet due to a bug in `core/io/zip_io.cpp`. The `zipio_open` function doesn't properly handle the `ZLIB_FILEFUNC_MODE_EXISTING` flag. I'll fix that in another PR.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/1069.*

### CODE REVIEW

**Comment by Calinou** [comment]
It seems the rebase failed, as lots of unrelated commits are included.

**Comment by flyingpimonster** [comment]
Ah, oops, I rebased on 3.2 instead of master.

I've renamed the classes and fixed a couple build errors.

**Comment by avencherus** [comment]
@akien-mga Is this still a feature you'd want to include in some version of 3.2?

**Comment by Frontrider** [comment]
What is the status on this?

**Comment by Calinou** [comment]
@Frontrider The `master` branch is currently undergoing refactoring, which means it can take a while for pull requests to be merged (especially if they introduce large features like this one).

**Comment by avril-gh** [comment]
@marcelofg55 

i tested this [zip-module-3.2.1](https://github.com/flyingpimonster/godot/tree/zip-module-3.2.1) , works nice, and its very useful,
would love to see it in 3.2 but you pushed to master instead ?

**Comment by asheraryam** [comment]
@avril-gh All new features are added to `master` first and then they're cherry-picked for older branches if it's stable enough and doesn't majorly break compatibility.

**Comment by avril-gh** [comment]
@asheraryam

Thank's for clarifying. Didnt know about this workflow. 👍 

> then they're cherry-picked for older branches if the changes are stable enough and don't majorly break compatibility.

been wondering, because i already tried it and this PR does not work with 3.2 due to differences in 4.0 . (while [zip-module-3.2.1](https://github.com/flyingpimonster/godot/tree/zip-module-3.2.1) do works) ...

_Oh well, whatever will be, it's a module so can have it even if its not merged._ :heart_eyes_cat:

**Comment by Calinou** [comment]
@avril-gh This pull request could be remade for the `3.2` branch if there's enough demand.

**Comment by flyingpimonster** [comment]
I've backported this to 3.2 in my branch here: https://github.com/flyingpimonster/godot/tree/zip-module-3.2 Should I open a separate pull request?

**Comment by Calinou** [comment]
@flyingpimonster Thanks for taking the time to backport it!

I think it would be better to have this one merged first, so we can see if there's anything to change before we can merge the `3.2` variant.

**Comment by Sungray** [comment]
I'm trying [zip-module-3.2.1](https://github.com/flyingpimonster/godot/tree/zip-module-3.2.1), works great except for the ZIPPacker.APPEND_ADDINZIP open mode. The zip file is being overwritten.

Open ZIPPACKER, write
a/nice/file.txt
Close

Open ZIPPACKER ZIPPacker.APPEND_ADDINZIP, write
a/not/nice/file.txt
Close

I only have a/not/nice/file.txt in the zip.

**Comment by Two-Tone** [comment]
Why does trying to open an already open zip close it? That doesn't make much sense to me as I'd expect it to just not do anything.  I'd also have to keep track of my zip files and their open/closed state just to make sure I don't accidentally close the zip at some point.

**Comment by flyingpimonster** [comment]
@Sungray, your issue seems to be caused by improper translation of file mode flags in [core/io/zip_io.cpp](https://github.com/godotengine/godot/blob/master/core/io/zip_io.cpp#L41). I'll push a commit to fix that.

**Comment by flyingpimonster** [comment]
@Two-Tone That's so you can reuse a ZIPPacker object by opening a different file--it will automatically close the old one and start using the new one. This is similar to how PCKPacker works.

**Comment by Sungray** [comment]
> 
> 
> @Sungray, your issue seems to be caused by improper translation of file mode flags in [core/io/zip_io.cpp](https://github.com/godotengine/godot/blob/master/core/io/zip_io.cpp#L41). I'll push a commit to fix that.

Yes, now it's working. Thanks a lot !

**Comment by cameronsteele** [comment]
This just saved me a bunch of time but for needing to download and compile the source. Many thanks to the programmer. Seems like a useful thing to merge into 3.2 since the work is already done. Moving to 4.0 already requires renaming instances of `yield`, and this feature could be put into the hands of people who aren't comfortable compiling from scratch today.

**Comment by LeonardMeagher2** [comment]
Is there a reason this isn't just a single class with methods of reading and writing?

**Comment by flyingpimonster** [comment]
> Is there a reason this isn't just a single class with methods of reading and writing?

IIRC that's how the minizip API works--there are separate open/close functions for reading vs. writing a zip file.

**Comment by flyingpimonster** [comment]
I've rebased on master and added examples to the docs. Is there anything else I need to do?

**Comment by akien-mga** [comment]
Thanks for the rebase and the docs! This mostly needs some of the core devs to review and test the code, but otherwise the feature is good IMO. We'll let you know if any changes are needed.

**Comment by avencherus** [comment]
> > Is there a reason this isn't just a single class with methods of reading and writing?
> 
> IIRC that's how the minizip API works--there are separate open/close functions for reading vs. writing a zip file.

Is there a functional reason it cannot be combined within the API?

**Comment by Frontrider** [comment]
> > > Is there a reason this isn't just a single class with methods of reading and writing?
> > 
> > 
> > IIRC that's how the minizip API works--there are separate open/close functions for reading vs. writing a zip file.
> 
> Is there a functional reason it cannot be combined within the API?

That is what I'm thinking.

**Comment by Two-Tone** [comment]
What about the unintuitive behavior of [ZIPReader closing a file if you try and open it twice](https://github.com/godotengine/godot/pull/34444/files#diff-99defdabb2b881658f72cb31b9fc1f1d2f4d8c4a394883d8e6b01c3ce668ace9R24)?  I can see people doing that by accident and creating bugs for themselves because of that unintuitive nature. I can't see any reason to close a file if you try and open it a second time.

**Comment by cameronsteele** [comment]
> 
> 
> What about the unintuitive behavior of [ZIPReader closing a file if you try and open it twice](https://github.com/godotengine/godot/pull/34444/files#diff-99defdabb2b881658f72cb31b9fc1f1d2f4d8c4a394883d8e6b01c3ce668ace9R24)? I can see people doing that by accident and creating bugs for themselves because of that unintuitive nature. I can't see any reason to close a file if you try and open it a second time.

@Two-Tone I believe the wording you linked refers to the fact that each instance of the class can only have one file open for reading at a time. If you open any second file, it will close the first one. If you reopen the same file, it should close the file and then open it again unless you have observed it functioning differently.

**Comment by Two-Tone** [comment]
Ahhh, I see. The wording isn't 100% clear and to me it reads as I describe. That line in the documentation should to be changed to ensure the wording is completely clear.

**Comment by flyingpimonster** [comment]
> > > Is there a reason this isn't just a single class with methods of reading and writing?
> > 
> > 
> > IIRC that's how the minizip API works--there are separate open/close functions for reading vs. writing a zip file.
> 
> Is there a functional reason it cannot be combined within the API?

Minizip has two separate APIs, one for reading files and one for writing them (see [unzip.h](https://github.com/godotengine/godot/blob/master/thirdparty/minizip/unzip.h) and [zip.h](https://github.com/godotengine/godot/blob/master/thirdparty/minizip/zip.h). To implement read/write on the same object, I'd have to reopen the file every time you switch. Even then, AFAICT minizip has no way to remove files from an archive.

The best way to use ZIPReader/ZIPPacker is to read an archive, modify the contents, then write a new archive.

**Review by Summersay415 — APPROVED** [review]

**Comment by MikeSchulze** [comment]
When will zip read/write support be included in the 3.x branch?
I just wonder why such a frequently requested feature takes so long to get into the Godot 3.x branch.

**Comment by akien-mga** [comment]
@MikeSchulze Same reason as explained here: https://github.com/godotengine/godot/pull/34444#issuecomment-745129943

**Comment by MikeSchulze** [comment]
> @MikeSchulze Same reason as explained here: [#34444 (comment)](https://github.com/godotengine/godot/pull/34444#issuecomment-745129943)
> 
> There's 1,200 PRs of backlog, it takes time for experienced contributors to process everything and have time to review things properly.

Thanks for the answer, 1.200 thats a lot ;)
I was just wondering because the topic is older more than one and a half year.

**Comment by flyingpimonster** [comment]
Rebased.

**Comment by qrrk** [comment]
Very much looking forward to this, as well as the backport to 3.x. It would untie my hands for a lot of optimizations and improvements in [my project](https://github.com/qrrk/Catapult).

I know it is not likely, but if there is anything I could do to expedite this PR, please let me know.

**Comment by mojoyup** [comment]
How would I access these two classes in a script? I am trying to use the filedialog to save a file as a .zip and also be able to open/edit images within the .zip for resaving using the signals if I can.

**Comment by akien-mga** [comment]
Sorry for the delay reviewing this - we finally discussed and approved the linked proposal, and this seems to be a fairly straightforward implementation for it.

Would be good if @bruvzg could review. Some changes will be needed to adapt to the new RefCounted `FileAccess` (see #59440). @Faless also had some comments about further work that might be needed to have file streaming for more elaborate use cases (but that's not blocking for this PR and is a general issue with Godot's file APIs).

Finally, I think that this would make sense to have in `core/io` instead of a module. The implementation is pretty straightforward and we already have core ZIP functionality - the only difference is that this exposes it in a user accessible manner.

**Comment by anderlli0053** [comment]
Can this be cherry-picked for Godot 3.5 or at least 3.x ?

**Comment by Calinou** [comment]
> Can this be cherry-picked for Godot 3.5 or at least 3.x ?

3.5 is nearing release, so I'm afraid not. There's [another pull request remaking this feature for `3.x`](https://github.com/godotengine/godot/pull/58900), but it'll have to wait for 3.6 to land in a sta	ble release.

**Comment by anderlli0053** [comment]
> > Can this be cherry-picked for Godot 3.5 or at least 3.x ?
> 
> 3.5 is nearing release, so I'm afraid not. There's [another pull request remaking this feature for `3.x`](https://github.com/godotengine/godot/pull/58900), but it'll have to wait for 3.6 to land in a stable release.

Thank you Hugo

File: modules/zip/zip_reader.cpp
Comment (touilleMan) [inline modules/zip/zip_reader.cpp:62]: Are we sure about this limit ?

filename contains not only the name of the file, but also it path within the archive (e.g. `foo/bar/spam.txt`)

So the limit is fine on Windows where path are limited to 250 characters (but I think to recall there is special case in Windows where this might be longer), but on Linux/Mac 255 is the max size of file name (and entire path is at least 1024 bytes).

So I think it would be better to set a bigger buffer here

Typically doing something like
```suggestion
		char filename[256];  // Note filename is a path !
		int err = unzGetCurrentFileInfo64(uzf, NULL, filename, sizeof(filename), NULL, 0, NULL, 0);
		if (err == UNZ_OK) {
			s.push_back(filename);
		} else {
			// Assume filename buffer was too small
			unz_file_info64 file_info;
			int err = unzGetCurrentFileInfo64(uzf, &file_info, NULL, 0, NULL, 0, NULL, 0);
			char long_filename_buff  = memalloc(file_info.size_filename);
			int err = unzGetCurrentFileInfo64(uzf, NULL, long_filename_buff, sizeof(filename), NULL, 0, NULL, 0);
			s.push_back(long_filename_buff);
			memfree(long_filename_buff);
		}
```

Comment (bruvzg) [reply modules/zip/zip_reader.cpp:62]: > So the limit is fine on Windows where path are limited to 250 characters

By default, Windows `MAX_PATH` is 260 (but it's possible to disable the limit on Windows 10+). Unicode UNC (`\\?\` prefixed) paths are limited to 32K UTF-16 characters.

Comment (ssokolow) [reply modules/zip/zip_reader.cpp:62]: I'm not sure what limits Zip itself imposes, but:

1. If you use `\\?\`-format paths or enable Win32 long paths via group policy or the registry on Windows 10, then you're subject to much longer filesystem-based length limits on Windows.
2. Various POSIX-specified libc functions are limited to `PATH_MAX` (`limits.h`) by a buffer allocation, which is typically 4096, but you [can work around that](https://insanecoding.blogspot.com/2007/11/pathmax-simply-isnt.html).

Here are the limits I'm aware of:

* **FAT32:** 32,760 "Unicode characters" for paths, 255 "Unicode characters" for filenames
* **exFAT:** 32,760 "Unicode characters", 255 "Unicode characters" for filenames
* **NTFS:** 32768 "Unicode characters" , 255 "Unicode characters" for filenames
* **ReFS:** 32768 "Unicode characters", 255 "Unicode characters" for filenames
* **UDF:** 1023 bytes, , 255 bytes for filenames
* **ext2/3/4:** No limit for paths (As I demonstrate with some Python [here](https://askubuntu.com/questions/859945/what-is-the-maximum-length-of-a-file-path-in-ubuntu/1169977#1169977)), 255 bytes for filenames
* **HFS+:** No limit for paths as far as I was able to google up, 255 bytes for filenames

As far as I'm aware, "Unicode characters" means 16-bit code units.

Comment (bruvzg) [reply modules/zip/zip_reader.cpp:62]: ZIP format uses a 16-bit value to store filename length, so it's probably limited to 64K (UTF-8). Not sure if there are extra limitations, but it's probably better to always use `unz_file_info64` and allocate space for filename on demand.

Comment (ssokolow) [reply modules/zip/zip_reader.cpp:62]: Also...

> So the limit is fine on Windows where path are limited to 250 characters

Actually, if I remember correctly, the limit is **260** characters and they chose that limit based on `A:\<some buffer that allocates for 256 chars><NUL>`.

> ZIP format uses a 16-bit value to store filename length, so it's probably limited to 64K (UTF-8).

Yeah. No clue about the upper limit, but there are plenty of google results with people complaining about this or that unzipping tool erroring out because the Zip file contains paths longer than 260 and the tool isn't using `\\?\`.

**Comment by touilleMan** [comment]
@akien-mga do you think this PR will be able to make it into Godot 4.0 ?

The asset store currently doesn't play well with Godot-Python (the assets must be stored as a git repo and there is no possibility to provide different asset per platform, so for Godot 3 I have to [put ~200mo of data on a git repo for each release...](https://github.com/touilleMan/godot-python-assetlib-repo/tree/master/addons/pythonscript) )

Long story short, I wrote a [Godot-Python installer addon](https://github.com/touilleMan/godot-python-install-manager) in GDScript that download just the right data. This works great except I need a way bundle data...

- I tried using the tar command present on the system but it's not reliable (not available on all versions of Windows)
- I currently use a pure GDScript implementation of zip, but it is not 100% reliable

So in the end having this PR merged for Godot 4.0 would help me a lot :smiley:  (of course Godot 4.0 feature freeze might already be there, so I would totally understand if this train as already left ^^)

**Comment by fire** [comment]
@RevoluPowered You mentioned you wanted to read and write zips.

**Comment by MisterMX** [comment]
I created a squashed version of this PR's branch and rebased it on the lastest master: https://github.com/MisterMX/godot/tree/zip-module

Not sure, if I should create a second PR or if @jameswestman wants to copy the changes.

**Comment by mhilbrunner** [comment]
@MisterMX You might want to integrate the file path size changes suggested by touilleMan above if you haven't already, and then open a new PR and reference this one in the description.

For the actual commit, you can still give credit to the original author by using [`Co-authored-by`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

Thanks for salvaging this!

**Comment by Anutrix** [comment]
Is this still needed after #65281?

**Comment by MisterMX** [comment]
Yes!

**Comment by mhilbrunner** [comment]
Closing as salvaged in https://github.com/godotengine/godot/pull/65281.

**Pull request closed without merging**
